src/utils/buildTaggedTemplate.js

```diff
@@ -0,0 +1,123 @@
+import { dirname, relative } from 'path';
+import { stripIndent } from 'common-tags';
+import groupBy from 'lodash/groupBy';
+import uniq from 'lodash/uniq';
+
+const rComposes = /\b(?:composes\s*?:\s*([^;>]*?)(?:from\s(.+?))?(?=[;}/\n\r]))/gim;
+const rPlaceholder = /###ASTROTURF_PLACEHOLDER_\d*?###/g;
+
+function resolveInterpolation(path, nodeMap, localStyle) {
+  let nextPath = path.resolve();
+  while (nextPath && nextPath.isMemberExpression()) {
+    nextPath = nextPath.get('object').resolve();
+  }
+  const style = nextPath && nodeMap.get(nextPath.node);
+  if (!style) return null;
+
+  return {
+    isStyledComponent: style.isStyledComponent,
+    externalName: !style.isStyledComponent
+      ? path.get('property').node.name
+      : style.name,
+    importName: relative(
+      dirname(localStyle.absoluteFilePath),
+      style.absoluteFilePath,
+    ),
+  };
+}
+
+let id = 0;
+const getPlaceholder = () => `###ASTROTURF_PLACEHOLDER_${id++}###`;
+
+export default (path, nodeMap, localStyle, { tagName }) => {
+  const { quasi } = path.node;
+
+  const interpolations = new Map();
+  const expressions = path.get('quasi.expressions');
+
+  let text = '';
+  quasi.quasis.forEach((tmplNode, idx) => {
+    const { cooked } = tmplNode.value;
+    const expr = expressions[idx];
+
+    text += cooked;
+
+    if (expr) {
+      const result = expr.evaluate();
+
+      if (result.confident) {
+        text += result.value;
+        return;
+      }
+
+      // TODO: dedupe the same expressions in a tag
+      const interpolation = resolveInterpolation(expr, nodeMap, localStyle);
+
+      if (!interpolation) {
+        throw expr.buildCodeFrameError(
+          `Could not resolve interpolation to a value, ${tagName} returned class name, or styled component. ` +
```

Comment:
probably just use a long string? it's not like we have linting that complains and forces you to use the `str1 + str2` pattern

Reply:
I like being able to read the whole thing without scrolling tho